stor/dx.py

```diff
@@ -113,96 +165,727 @@ def temp_url(self, lifetime=300, filename=None):
+        For canonical paths like
+        DXPath('dx://project-FJq288j0GPJbFJZX43fV5YP1:file-FKzjvV80jfv8pFzQ0KfbYY6B'),
+        the resource is PosixPath('/file-FKzjvV80jfv8pFzQ0KfbYY6B').
+
+        The operation parses the object string, and hence no API request is performed.
+        """
         parts = self._get_parts()
         joined_resource = '/'.join(parts[1:]) if len(parts) > 1 else None
         return self.parts_class('/'+joined_resource) if joined_resource else None
 
     def dirname(self):
+        """Returns directory name of path. Returns self if path is a project.
+
+        For canonical paths, returns the project ID, since this parses the object
+        as string without making API calls to the server.
+        """
         if not self.resource:
             return self
         else:
-            return super(DXPath, self).dirname()
+            parts = self._get_parts()
+            if len(parts) == 2:  # paths like ('dx://proj:file') need different logic
+                parts[0] += ':'
+                new_path = '/'.join(parts)
+                return self.path_class(self.drive + self.path_module.dirname(new_path))
+            else:
+                return super(DXPath, self).dirname()
+
+    @property
+    def name(self):
+        """Returns File or folder name of the path. Empty string for projects or folders
+        with trailing slash.
 
-    def download_objects(self):  # may not need it
-        raise NotImplementedError
+        Makes no API calls to server, canonical paths are treated normally, and the basename
+        of the path is returned.
+        """
+        parts = self._get_parts()
+        if len(parts) == 2 and parts[1]:  # paths like ('dx://proj:file') need different logic
+            parts[0] += ':'
+            new_path = '/'.join(parts)
+            return self.parts_class(self.path_module.basename(new_path))
+        else:
+            return super(DXPath, self).name
 
     def remove(self):
-        raise NotImplementedError
+        """Removes a single object from DX platform
 
+        Raises:
+            ValueError: The path is invalid.
+        """
+        if not self.resource:
+            raise ValueError('DXPath must point to single data object to call remove')
+        file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                   project=self.canonical_project)
+        with _propagate_dx_exceptions():
+            file_handler.remove()
+        self.clear_cached_properties()
+
+    @_propagate_dx_exceptions()
     def rmtree(self):
-        raise NotImplementedError
+        """
+        Removes a resource and all of its contents.
+        The path should point to a project or directory.
+
+        Raises:
+            NotFoundError: The path points to a nonexistent directory
+        """
+        proj_handler = dxpy.DXProject(self.canonical_project)
+        if not self.resource:
+            return proj_handler.destroy()
```

Comment:
I thought we weren't going to allow folks to delete projects?

Reply:
Do you want me to raise an error here or listdir on root folder, and recursively call rmtree on folders and remove on files (equivalent to deleting root folder)?

Reply:
Waiting on your answer for this one...

Reply:
This sounds good:

```
listdir on root folder, and iteratively call rmtree on folders and remove on files (equivalent to deleting root folder)
```

Reply:
Added functionality and test case